### ISSUE TITLE
parser: redeclare variable declaration with function declaration in switch CaseBlock in non-strict mode should throw a syntax error

### ISSUE THREAD

**Issue by Dunqing** [oxc-project/oxc#10050]
[Test262 test]( https://github.com/tc39/test262/blob/bc5c14176e2b11a78859571eb693f028c8822458/test/language/statements/switch/syntax/redeclaration/var-name-redeclaration-attempt-with-function.js) | [Babel Playground](https://babeljs.io/repl#?browsers=defaults%2C%20not%20ie%2011%2C%20not%20ie_mob%2011&build=&builtIns=false&corejs=3.21&spec=false&loose=false&code_lz=M4dwlgLgxgFgBACgAwEo4G85QIbAKZwCMAXHAG7YBOcAZgNxwAmeN2ArgDYSk1sB2UCGAD2fWgjToAvnCkAoIA&debug=false&forceAllTransforms=false&modules=false&shippedProposals=false&evaluate=false&fileSize=false&timeTravel=false&sourceType=script&lineWrap=true&presets=&prettier=true&targets=&version=7.27.0&externalPlugins=%40babel%2Fplugin-transform-object-rest-spread%407.25.9&assumptions=%7B%7D) | [OXC Playground](https://playground.oxc.rs/#eNo9kM1OwzAMgF8l8mlIE9qQuJQr2hUEHLl4qTOC0qSynf1o6rvjtrSn2Irt77Pv4KEBuUT1P26ze3B351HI7Rt3RnbhxbUUsCZtXKjZayzZhY2VDW74zrCFAs0duObxkVtWvEKjXGkLKWZdYvGlpzW5dceSlkwZs4TCHTQBk9CwhR5ZiMeJmFK5fJBWzm9VJbZ0+LdY2nsmqz3TOzJlWRGlsqdDTJSxMzAoiT7+ytWM57+v2yhkYhx7BYOOuhPU4tVptlDkE9kuQPK02z/bjCgloVL7Sj4h4yhk7MXfl5ZONJ3E+MdEnxOyw372s5Iu5hjiwvMlK5d0sGXHpjPxsYjpzQOH4Q+qfpVB)

```js
// This file was procedurally generated from the following sources:
// - src/declarations/function.case
// - src/declarations/redeclare-allow-sloppy-function/switch-attempt-to-redeclare-var-declaration.template
/*---
description: redeclaration with FunctionDeclaration (VariableDeclaration in SwitchStatement)
esid: sec-switch-statement-static-semantics-early-errors
flags: [generated]
negative:
  phase: parse
  type: SyntaxError
info: |
    SwitchStatement : switch ( Expression ) CaseBlock

    It is a Syntax Error if any element of the LexicallyDeclaredNames of
    CaseBlock also occurs in the VarDeclaredNames of CaseBlock.

---*/


$DONOTEVALUATE();

switch (0) { case 1: var f; default: function f() {} }

```

### Expect

Both should throw a syntax error in non-strict mode and strict mode

### Actual

Non-strict mode has no error, and strict mode has a syntax error
